zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -343,17 +343,54 @@ public void subscribeDataChanges(String path, IZkDataListener listener) {
    * Subscribe RecursivePersistListener for a particular path. User can only subscribe when
    * `_usePersistWatcher` is set to true and there is no pre-existing watcher on the path.
    */
-  // TODO: Add impl and remove exception
-  public boolean subscribePersistRecursiveWatcher(String path,
-      RecursivePersistListener recursivePersistListener)
-      throws KeeperException.UnimplementedException {
-    throw new KeeperException.UnimplementedException();
+  public void subscribePersistRecursiveListener(String path,
+      RecursivePersistListener recursivePersistListener) {
+    if (!_usePersistWatcher) {
```

Comment:
this does throw exception, in that case, should the higher level call return 'false' or we need exception to pass through?

Reply:
TFTR! 
We do return `false` in ZkMetaclient when it is in false path (try to subscribe a non exist path and flag for "skipWatchingNonExistNode" is true). Meaning there is no unexpected behavior.
 When `subscribePersistRecursiveListener ` throws exception, it means native ZK is unable to process, witch is categorized as an exception, so we throw it to users.